ring-core/src/ring/middleware/multipart_params.clj

```diff
@@ -55,23 +56,44 @@
     (file-item-iterator-seq
       (.getItemIterator ^FileUpload upload context))))
 
+(defn- parse-content-type-charset [^FileItemStream item]
+  (some->> (.getContentType item)
+           (re-find re-charset)
+           second))
+
+(defn- parse-html5-charset [params]
+  (when-let [charset (->> params (filter #(= (first %) "_charset_")) first second :bytes)]
+    (String. ^bytes charset "US-ASCII")))
+
+(defn- decode-string-values [fallback-encoding forced-encoding params]
+  (let [html5-enc (parse-html5-charset params)]
+    (map
```

Comment:
Rather than use `map` here, a `for` might be more concise:

```clojure
(let [html5-encoding (parse-html5-charset params)]
  (for [[k v] params]
    [k (if-let [^bytes bytes (:bytes v)]
         (String. bytes (or forced-encoding
                            html5-encoding
                            (:item-encoding v)
                            fallback-encoding)
         v)]))